File: 0303. Range Sum Query - Immutable/0303.java
/**
 * Time Complexity:
 * - Constructor: O(1) → Just assigns the reference to the input array.
 * - sumRange(left, right): O(n) → Iterates through the entire array to check indices.
 *
 * Space Complexity:
 * - O(1) → No extra space is used apart from storing the reference to nums.
 */
class NumArray {
    private int[] nums; // Stores the input array

    // Constructor to initialize the nums array
    public NumArray(int[] nums) {
        this.nums = nums;
    }
    
    public int sumRange(int left, int right) {
        int total=0;
        // Iterate through the entire array and sum only the relevant range
        for(int i=0; i<nums.length; i++){
            if(left <= i && i<=right){
                total+=nums[i];
            }
        }
        return total;
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(left,right);
 */

File: 0303. Range Sum Query - Immutable/0303.py
# time complexity: O(n)
# space complexity: O(1)
class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self._nums = nums

    def sumRange(self, left, right):
        """
        :type left: int
        :type right: int
        :rtype: int
        """
        result = 0
        for i in range(len(self._nums)):
            if left <= i and i <= right:
                result += self._nums[i]
        return result


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)